File: Reference Implementation using python/modules/analyzer.py
import math
import os

def calculate_entropy(file_path):
    """
    Reads a file in binary mode and calculates its Shannon Entropy.
    Returns a float between 0.0 (no randomness) and 8.0 (total randomness/encrypted).
    """
    try:
        # Check if file is empty to avoid division by zero
        file_size = os.path.getsize(file_path)
        if file_size == 0:
            return 0.0

        # Read the file's raw bytes
        with open(file_path, 'rb') as f:
            byte_data = f.read()

        # Create an array to count the occurrence of each of the 256 possible bytes
        byte_counts = [0] * 256
        for byte in byte_data:
            byte_counts[byte] += 1

        # Apply the Shannon Entropy Formula: H(X) = -sum(P(x) * log2(P(x)))
        entropy = 0.0
        for count in byte_counts:
            if count > 0:
                probability = count / file_size
                entropy -= probability * math.log2(probability)

        return entropy

    except Exception as e:
        print(f"      [ERR] Analyzer failed to read file: {e}")
        return 0.0

File: src/analyzer.cpp
#include "analyzer.hpp"
#include <fstream>
#include <vector>
#include <cmath>
#include <iostream>

double calculate_entropy(const std::string& file_path) {
    // Open file in binary mode, starting at the end to get the size
    std::ifstream file(file_path, std::ios::binary | std::ios::ate);
    if (!file.is_open()) {
        std::cerr << "      [ERR] Analyzer failed to open file.\n";
        return 0.0;
    }

    std::streamsize size = file.tellg();
    file.seekg(0, std::ios::beg); // Rewind to the beginning

    if (size == 0) return 0.0;

    // Read the entire file into a high-speed buffer
    std::vector<unsigned char> buffer(size);
    if (!file.read(reinterpret_cast<char*>(buffer.data()), size)) {
        return 0.0;
    }

    // Array to hold the frequency of all 256 possible bytes
    std::vector<int> byte_counts(256, 0);
    for (unsigned char byte : buffer) {
        byte_counts[byte]++;
    }

    // Shannon Entropy Formula: H(X) = -sum(P(x) * log2(P(x)))
    double entropy = 0.0;
    for (int count : byte_counts) {
        if (count > 0) {
            double probability = static_cast<double>(count) / size;
            entropy -= probability * std::log2(probability);
        }
    }

    return entropy;
}
